src/frontend/flight/sync.rs

```diff
@@ -0,0 +1,378 @@
+    ) -> Result<(Option<SequenceNumber>, Option<SequenceNumber>)> {
+        let url = log_store.root_uri();
+
+        let orseq = (origin.clone(), sequence_number);
+
+        // Upsert a sequence entry for this origin/LSN
+        self
+            .seqs
+            .entry(orseq.clone())
+            .and_modify(|seq| {
+                if !seq.locs.contains(&url) {
+                    debug!("Adding {url} as sync destination for {origin}, {sequence_number}");
+                    seq.locs.insert(url.clone());
+                }
+
+                if last {
+                    debug!(
+                        "Received last sync for {url} from {origin}, {sequence_number}"
+                    );
+                    seq.last = true;
+                }
+            })
+            .or_insert(DataSyncSequence {
+                last,
+                locs: HashSet::from([url.clone()]),
+            });
+
+        // Finally upsert the new sync item for this location
+        let size = batches
+            .iter()
+            .fold(0, |bytes, batch| bytes + batch.get_array_memory_size());
+
+        let item = DataSyncItem {
+            origin_sequence: orseq,
+            pk_columns,
+            batches,
+        };
+        self.syncs
+            .entry(url)
+            .and_modify(|entry| {
+                entry.syncs.push(item.clone());
+                entry.size += size;
+            })
+            .or_insert(DataSyncCollection {
+                size,
+                insertion_time: now(),
+                log_store,
+                syncs: vec![item],
+            });
+
+        // Update the total size
+        self.size += size;
+
+        // Flag the sequence as volatile persisted for this origin if it is the last sync command
+        if last {
+            self.origin_memory.insert(origin.clone(), sequence_number);
+        }
+
+        self.flush_batches().await?;
+
+        Ok(self.stored_sequences(&origin))
+    }
+
+    // Criteria for return the cached entry ready to be persisted to storage.
+    // First flush any records that are explicitly beyond the configured max
+    // lag, followed by further entries if we're still above max cache size.
+    fn flush_ready(&mut self) -> Option<String> {
+        if let Some((url, sync)) = self.syncs.first()
+            && now() - sync.insertion_time
+                >= self.context.config.misc.sync_data.max_replication_lag_s
+        {
+            // First flush any changes that are past the configured max duration
+            // TODO: do this out-of-band
+            Some(url.clone())
+        } else if self.size >= self.context.config.misc.sync_data.max_in_memory_bytes {
+            // Or if we're over the size limit flush the oldest entry
+            Some(self.syncs.first().unwrap().0.clone())
+        } else {
+            None
+        }
+    }
+
+    async fn flush_batches(&mut self) -> Result<()> {
+        while let Some(url) = self.flush_ready() {
```

Comment:
The fact that we flush (and commit LSNs to Delta) url-by-url still breaks my brain - I need to think more about this tomorrow.

Say I have:

```
insert(table_1, [item_1])
insert(table_2, [item_2])
modify(table_2, [add item_3])
insert(table_3, [item_4]))
modify(table_2, [add item_5])
modify(table_3, [add item_6])
```

(and each item is a single-batch transaction, i.e. last=true always)

does the queue look like this?

```
[highest priority item, first to flush]
1. table_1: item_1
2. table_2: item_2, item_3, item_5
3. table_3: item_4, item_6
```

so if we flush only the first two items, will we say that we're durable up to 5, which is not true because we haven't flushed item_4 from table_3 yet?

Reply:
You're right about the state of the queue, and that this is a potential problem, will think about ways to resolve.

Reply:
One potential resolution is outlined in 666a70dadc84f38f6733e68b10a5cf33049c2636. It's not simple but it's a first pass and can be simplified further if we decide to go along with this approach.